lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -323,14 +352,24 @@ def from_requirement_dict(cls, collection_req, art_mgr, validate_signature_optio
                 req_signature_sources = [req_signature_sources]
             req_signature_sources = frozenset(req_signature_sources)
 
+        if (
+            req_version and req_type == 'subdirs'
+        ):
+            raise AnsibleError(
+                f"The {req_name} 'version' key is not applicable when 'type' is set to 'subdirs'."
+            )
+
+        if (
+            req_name is not None
+            and not AnsibleCollectionRef.is_valid_collection_name(req_name)
+            and req_source is not None
+        ):
+            raise AnsibleError(
+                f"The {req_name} 'source' key is not applicable when 'name' is not a FQCN."
+            )
+
         if req_type is None:
-            if (  # FIXME: decide on the future behavior:
-                    _ALLOW_CONCRETE_POINTER_IN_SOURCE
```

Comment:
My interpretation of this block is that this checks if the `source` key points to an artifact. Since the desired behavior is that the `source` key shouldn't point to an artifact, I thought this block was no longer necessary. Perhaps I'm wrong and this should stay in?

Reply:
I don't remember why this was added. Inspecting https://github.com/ansible/ansible/commit/595413d11346b6f26bb3d9df2d8e05f2747508a3 / https://github.com/ansible/ansible/pull/72591 does not shed any light on that. Perhaps @s-hertel, @jborean93 and I talked about it on our weekly calls. But I doubt they'll remember discussions that were taking place over 4 years ago.

But looking at it now, I'd assume that we didn't want people to put in ambiguous things into this field without specifying the type explicitly. And since there's this always-false feature flag `_ALLOW_CONCRETE_POINTER_IN_SOURCE`, it was probably something that might've been seen as controversial at the time.

This check seems to prevent assuming the `galaxy` type (the following `elif` branch) if something looks like a path to an artifact.